IPC implants are now mechfab-only

pwbokie opened the pull request:
<!-- By ticking or leaving ticked the option "Allow edits and access to secrets by maintainers" you give permission for repository maintainers to push changes to your branch without explicitly asking. -->

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do

<!-- Include a small to medium description of what your PR changes. -->
<!-- Document all changes, as not doing this may delay reviews or even discourage maintainers from merging your PR! -->
<!-- If your PR fixes an issue, add "Fixes #1234" somewhere in the PR description. This will automatically close the bug upon PR submission. -->
- Makes it so only the Mechfab can print IPC implants. Protolathes can't anymore.
- Restructures the Mechfab a bit, moving IPC parts to their own new section, IPC implants to their own new section called IPC Upgrades, and renaming the Cyborg Upgrade Modules section to Cyborg Upgrades (and moving it to be next to the other Cyborg sections). Just makes it all consistent and nicer.

## Why It's Good For The Game

<!-- Add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
Working on IPCs should be a stronger part of the Roboticist identity. Officially putting IPC parts and implants strongly into the mechfab forwards this goal.

## Testing

<!-- How did you test the PR, if at all? -->
- Scrolled through the mechfab to make sure it looked right.
- Checked the protolathe to make sure IPC implants weren't present.

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/docs/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  <!-- Replace the box with [x] to mark as complete. -->
  <!-- Ensure there are no spaces between the x and the square brackets [] else this will not work properly. -->
<hr>

![image](https://github.com/user-attachments/assets/40967a70-09a0-4115-9a6d-33520aed6646)


## Changelog

:cl:
tweak: IPC implants are now mechfab-only
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

1080pCat commented:
> * Makes it so only the Mechfab can print IPC implants. R&D can't anymore.
> 
> * Restructures the Mechfab a bit, moving IPC parts to their own new section, IPC implants to their own new section called IPC Upgrades, and renaming the Cyborg Upgrade Modules section to Cyborg Upgrades (and moving it to be next to the other Cyborg sections). Just makes it all consistent and nicer.

this should *probably* be an atomic pr

Ahzpy commented:
does this mean ipcs can now ingest pills? lord save us from the chemgaming

pwbokie commented:
I accidentally closed this by messing around with my fork. Oops.

pwbokie commented:
> > * Makes it so only the Mechfab can print IPC implants. R&D can't anymore.
> > * Restructures the Mechfab a bit, moving IPC parts to their own new section, IPC implants to their own new section called IPC Upgrades, and renaming the Cyborg Upgrade Modules section to Cyborg Upgrades (and moving it to be next to the other Cyborg sections). Just makes it all consistent and nicer.
> 
> this should _probably_ be an atomic pr

Agreed. I'll split it up if it gets feedback any more complex than a simple go-ahead.

Burzah reviewed (commented):
Can you update your Change Log for the new implant you are adding? It should be a bit more descriptive.

pwbokie commented:
> Can you update your Change Log for the new implant you are adding? It should be a bit more descriptive.

> add: A new culinary IPC implant which lets them eat and taste.

Done

Drsmail commented on code/modules/mob/living/taste.dm:
```diff
@@ -2,10 +2,11 @@
 		return TASTE_SENSITIVITY_NORMAL
+	else if(dna.species)
+		return dna.species.taste_sensitivity
+	return TASTE_SENSITIVITY_NORMAL
```
This looks odd to me. You sure you need to return this in last case?

Drsmail commented on code/modules/surgery/organs/augments_internal.dm:
```diff
@@ -947,10 +963,10 @@
 /obj/item/organ/internal/cyberimp/chest/ipc_joints/magnetic_joints/insert(mob/living/carbon/M, special = FALSE)
 	..()
-	ADD_TRAIT(M, TRAIT_IPC_JOINTS_MAG, "ipc_joint[UID()]")
+	ADD_TRAIT(M, TRAIT_IPC_CAN_EAT, "ipc_can_eat[UID()]")
```
This should be a mistake, right?

pwbokie replied:
yep, messed up a ctrl+F at some point apparently and didn't notice. thanks

Drsmail commented on code/modules/surgery/organs/augments_internal.dm:
```diff
@@ -947,10 +963,10 @@
+	ADD_TRAIT(M, TRAIT_IPC_CAN_EAT, "ipc_can_eat[UID()]")
 
 /obj/item/organ/internal/cyberimp/chest/ipc_joints/magnetic_joints/remove(mob/living/carbon/M, special = FALSE)
-	REMOVE_TRAIT(M, TRAIT_IPC_JOINTS_MAG, "ipc_joint[UID()]")
```
Was it changed unintentionally?

Drsmail reviewed (commented):
Looks good to me, just have a questions with ipc joints traits.

Drsmail reviewed (approved):
LGTM

Denthamos commented:
On another note
Adding a new implant AND making them all be roboticist only via the mechfab should of been two different PR's as well.  The mechfab change is a big change up in an aspect of science and should of been judged/voted on as a separate matter from the new implant being added.

That being said, I am not against the change and the addition.

pwbokie commented:
> On another note Adding a new implant AND making them all be roboticist only via the mechfab should of been two different PR's as well. The mechfab change is a big change up in an aspect of science and should of been judged/voted on as a separate matter from the new implant being added. As per our github policy and rules
> 
> > Your pull requests should be atomic. Balance changes, bug fixes and new features should all be in separate pull requests.
> 
> That being said, I am not against the change and the addition.

Agreed and I addressed this above. I can close this and split it up if needed, and I'll do this correctly in the future.

PollardTheDragon commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -1135,7 +1135,7 @@
 	category = list("Exosuit Equipment")
 
-//Cyborg Upgrade Modules
+//Cyborg Upgrades
```
```suggestion
// Cyborg Upgrades
```

PollardTheDragon commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -1395,6 +1377,83 @@
 	construction_time = 20 SECONDS
+	category = list("IPC")
+
+//IPC Upgrades
```
```suggestion
// IPC Upgrades
```

PollardTheDragon commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -1395,6 +1377,83 @@
+	build_path = /obj/item/organ/internal/cyberimp/chest/ipc_joints/flayer_pacification
+	category = list("IPC Upgrades")
+
+//Misc
```
```suggestion
// Misc
```

PollardTheDragon commented on code/modules/surgery/organs/augments_internal.dm:
```diff
@@ -896,6 +896,22 @@
+	desc = "This implant emulates the functions of a gastrointestinal system, allowing IPCs to eat and experience taste."
+	implant_color = "#0ac0d8"
+	origin_tech = "materials=2;powerstorage=2;biotech=2"
+	slot = "gastrointestinal" //Having a stomach implant with this is fine.
```
```suggestion
	slot = "gastrointestinal" // Having a stomach implant with this is fine.
```

PollardTheDragon commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -1395,6 +1377,83 @@
+
+//IPC Upgrades
+
+/datum/design/culinary_processing_implant
```
I do think this would be best as its own separate PR. One for the changes to improve the organization of the mechfab and move IPC parts to it, then another to add the culinary implant. This would keep the PRs atomic.

lewcc commented on code/modules/mob/living/carbon/human/human_mob.dm:
```diff
@@ -1671,19 +1671,21 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
-		to_chat(src, "Where do you intend to put \the [toEat]? You don't have a mouth!")
-		return FALSE
+		if(!ismachineperson(src) || (ismachineperson(src) && !HAS_TRAIT(src, TRAIT_IPC_CAN_EAT)))
+			to_chat(src, "Where do you intend to put \the [toEat]? You don't have a mouth!")
```
```suggestion
			to_chat(src, "Where do you intend to put [toEat]? You don't have a mouth!")
```

lewcc commented on code/modules/mob/living/carbon/human/human_mob.dm:
```diff
@@ -1671,19 +1671,21 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
-			return FALSE
+		if(!ismachineperson(src) || !HAS_TRAIT(src, TRAIT_IPC_CAN_EAT))
+			if(!((istype(toEat, /obj/item/reagent_containers/drinks) && (ismachineperson(src)))))
+				to_chat(user, "Where do you intend to put \the [toEat]? \The [src] doesn't have a mouth!")
```
```suggestion
				to_chat(user, "Where do you intend to put [toEat]? \The [src] doesn't have a mouth!")
```

lewcc reviewed (changes requested):
Agreed with pollard: can you please break the new implant out to a new PR?

pwbokie commented:
Closing to atomize
